Make extension method resolution deterministic

Akirathan opened the pull request:
Fixes #1243

### Pull Request Description

Fixes the non-deterministic method resolution in some cases. Normally, method resolution tries to lookup the method in:
1. In the scope where the type is defined
2. In the current module scope
3. In imported module scopes

The third step used to be non-deterministic. I have made it deterministic simply by replacing `java.util.HashMap` by `java.util.LinkedHashMap` in [ModuleScope](https://github.com/enso-org/enso/blob/71d2e97faa5236a626830e5ed4d2e04f18b5e6c6/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java).

### Important Notes

I have attempted three times to implement this feature:
- In static compilation in a separate compiler pass
- Right after static compilation in `IrToTruffle`, just before execution.
- At runtime, just before method is executed in `UnresolvedSymbol`, as an error.
All these attempts failed. More details in [comment below](https://github.com/enso-org/enso/pull/9844#issuecomment-2122946274).

At the end, I have decided to fix this bug by **making the method resolution deterministic** and not throwing any errors.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

Commit:
Move createProject method to TestBase

Commit:
Add ExtensionMethodResolutionTest

Commit:
Make the test more general.

Now, it tests all the shadowing, not only extension methods.

Commit:
Shadowing is semantically correct - revert some tests

Commit:
rename errors.Redefined.Method.atomName to typeName

Commit:
Revert renaming of the test

Commit:
OverloadResolution compiler pass checks redefinition of static methods from imported modules.

Commit:
Fix a test - make sure extension method is imported

Commit:
fmt

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/ExtensionMethodResolutionTest.java:
```diff
@@ -0,0 +1,136 @@
+        polyCtx.getTopScope().compile(true);
+        fail("Expected compilation error: " + out);
+      } catch (PolyglotException e) {
+        assertThat(e.getMessage(), is(CompilationAbortedException.COMPILATION_ABORTED_MSG));
```
Recognizing exceptions by their message? Unusual. I'd rather hardcode the message string directly into the test....


I think you can `assertTrue(e.isSyntax())`...

Akirathan replied:
Refactored to `assertTrue(e.isSyntaxError())` in 3cd30ab464b200946fe37690e28a0ee1ac6122ff

JaroslavTulach reviewed (approved):
Looks like you managed to detect this error during static compilation. Good!

Commit:
Handle the case when current module imports itself.

Commit:
bindingMap can be null

Commit:
fmt

Commit:
Do not check for exception message in test

Commit:
Add project creation methods to TestBase

Commit:
projDir must be a directory

Commit:
Add test on same method in shadowed type.

Currently, it fails.

Commit:
IRDumper handles Name.MethodReference IR

Commit:
OverloadsResolution pass deals only with extension methods

Commit:
Fix iteration over extension methods only.

Commit:
fmt

Commit:
Add failing test

Inspired by code in Sieve_Without_Types.enso

Commit:
Restore OverloadsResolution to the version from develop

Commit:
IrToTruffle fails when methodDef tries to overwrite an existing extension method.

Commit:
Assert first execution failure in tests

Commit:
fmt

Commit:
tests expect runtime error, not a compilation error

Commit:
Add ImportResolutionOrderTest

Commit:
fmt

Commit:
Add more tests for import resolution order

Commit:
Small refactoring of test

Commit:
Test ImportExportResolution rather than just ImportResolution.

This test matches the behavior of Compiler's import export resolution more closely.

Commit:
ambiguous method is checked in UnresolvedSymbol

Commit:
Tests must call ambiguous method to bump into the error

Commit:
Revert changes in IrToTruffle

Akirathan commented:
All attempts to check extension method conflicts during compilation failed:
- `OverloadResolution` compiler pass does not have information from `org.enso.interpreter.runtime.scope.ModuleScope`, and without it, it is very difficult to query defined methods in imported modules.
- Checking this in `IrToTruffle` failed as well, because the imported modules there might not be compiled, i.e, their stage might not be `CompilationStage.AFTER_CODEGEN`. Initially, I suspected this is a bug in import/export resolution, that it orders the modules in non-topological order, so I have defined tests in https://github.com/enso-org/enso/blob/097f016741420cf921345f9cb143c6ae0e57f50e/engine/runtime-integration-tests/src/test/java/org/enso/compiler/ImportExportResolutionOrderTest.java, but it turns out that this is a "feature" not a bug. And changing the behavior of import export resolution is very tricky.

At the end, I have resolved to checks at runtime, inside [UnresolvedSymbol](f11ef4d11756362b6727bde3364c98175e3d7486).

<blockquote><img src="https://opengraph.githubassets.com/0aed28d432dd3bc7d4e1f2a9fd926bd0a035f4de92daa13c86e54d9d5c9fefc2/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/blob/097f016741420cf921345f9cb143c6ae0e57f50e/engine/runtime-integration-tests/src/test/java/org/enso/compiler/ImportExportResolutionOrderTest.java">enso/engine/runtime-integration-tests/src/test/java/org/enso/compiler/ImportExportResolutionOrderTest.java at 097f016741420cf921345f9cb143c6ae0e57f50e · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

Commit:
Remove ImportExportResolutionOrderTest.

Turns out that current ordering of modules is a "feature" not a "bug". Let's not dig in that.

Commit:
Revert UnresolvedSymbol and ModuleScope to develop

Commit:
Update tests - method resolution should be deterministic, no errors

Commit:
ModuleScope has LinkedHashMap and LinkedHashSet only.

So that the iteration order of the entries is deterministic.

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:
```diff
@@ -408,12 +407,12 @@ public void reset() {
   public ModuleScope withTypes(List<String> typeNames) {
-    Map<String, Object> polyglotSymbols = new HashMap<>(this.polyglotSymbols);
-    Map<String, Type> requestedTypes = new HashMap<>(this.types);
-    Map<Type, Map<String, Supplier<Function>>> methods = new ConcurrentHashMap<>();
```
cc @hubertp. Note that I am getting rid of `ConcurrentHashMap` here and no tests are failing. This should be a proof that `ModuleScope` is intended to be single-threaded. Note that it would even not be possible to use it from more threads, as, e.g., in `reset` method in https://github.com/enso-org/enso/pull/9844/files#diff-c3869665e0ec34703f8e3eea0464924c0e4260eab392dbccdc202ec1f0270379L399, a normal `HashMap` is reassigned to the `methods` field instead of `ConcurrentHashMap`.
